hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/HRegionServer.java

```diff
@@ -3864,21 +3848,36 @@ public void remoteProcedureComplete(long procId, Throwable error) {
   }
 
   void reportProcedureDone(ReportProcedureDoneRequest request) throws IOException {
-    RegionServerStatusService.BlockingInterface rss = rssStub;
-    for (;;) {
+    RegionServerStatusService.BlockingInterface rss = ensureRssStub();
+    try {
+      rss.reportProcedureDone(null, request);
+    } catch (ServiceException se) {
+      resetRssStubOnServiceException(rss);
+      throw ProtobufUtil.getRemoteException(se);
+    }
+  }
+
+  private RegionServerStatusService.BlockingInterface ensureRssStub() {
+    return ensureRssStub(false, Integer.MAX_VALUE);
+  }
+
+  private RegionServerStatusService.BlockingInterface ensureRssStub(
```

Comment:
In general I think checkClusterConn should always be true. The old implementation for reportProcedureDone seems incorrect, as we can not quit if the rs is shutting down. Maybe we should throw an IOException when the clusterConnection is closed in reportProcedureDone.